nimbus/src/enrollment.rs

```diff
@@ -20,28 +23,300 @@ pub enum EnrolledReason {
     OptIn,     // Explicit opt-in.
 }
 
+// These are types we use internally for managing non-enrollments.
+#[derive(Deserialize, Serialize, Debug, Clone, Hash, Eq, PartialEq)]
+pub enum NotEnrolledReason {
+    OptOut,      // The user opted-out of experiments before we ever got enrolled to this one.
+    NotSelected, // The evaluator bucketing did not choose us.
+    NotTargeted, // We are not being target for this experiment.
+    NotRunning,  // The experiment enrollment is paused.
+}
+
+// These are types we use internally for managing disqualifications.
+#[derive(Deserialize, Serialize, Debug, Clone, Hash, Eq, PartialEq)]
+pub enum DisqualifiedReason {
+    OptOut,    // The user opted-out from this experiment or experiments in general.
+    Targeting, // The targeting has changed for an experiment.
+}
+
 // Every experiment has an ExperimentEnrollment, even when we aren't enrolled.
 #[derive(Deserialize, Serialize, Debug, Clone, PartialEq)]
 pub struct ExperimentEnrollment {
     pub slug: String,
     pub status: EnrollmentStatus,
 }
 
+pub struct EnrollmentChangeEvent {
+    pub experiment_slug: String,
+    pub branch_slug: String,
+    pub enrollment_id: String,
+    pub reason: Option<String>,
+    pub change: EnrollmentChangeEventType,
+}
+
+impl EnrollmentChangeEvent {
+    pub(crate) fn new(
+        slug: &str,
+        enrollment_id: &str,
+        branch: &str,
+        reason: Option<&str>,
+        change: EnrollmentChangeEventType,
+    ) -> Self {
+        Self {
+            experiment_slug: slug.to_owned(),
+            branch_slug: branch.to_owned(),
+            reason: reason.map(|s| s.to_owned()),
+            enrollment_id: enrollment_id.to_owned(),
+            change,
+        }
+    }
+}
+
+pub enum EnrollmentChangeEventType {
+    Enrollment,
+    Disqualification,
+    Unenrollment,
+}
+
+impl ExperimentEnrollment {
```

Comment:
more a style question than a nit: it seems a bit odd to me that when the struct is in the same file that the impl doesn't immediately follow. I guess it might make sense if the "rule" as "all decls first, followed by impls", but that doesn't seem to be the case here either (ie, there's an `enum` following this impl)